### ISSUE TITLE
Document that plugs in `pipe_through` (as opposed to those in `pipeline`) are not wrapped in a try/rescue that reraises `Plug.Conn.WrapperError`

### ISSUE THREAD

**Issue by palm86** [phoenixframework/phoenix#2097]
# Environment

* Elixir version: 1.4.0
* Phoenix version: 1.2.1
* NodeJS version: 6.9.1
* NPM version: 3.10.8
* Operating system: Ubuntu

# Issue

Bodies of plugs listed with the `pipe_through` macro should be wrapped in a try rescue that reraises `Plug.Conn.WrapperError` or it should be documented that it does not. Instead, only bodies of plugs listed in the `pipeline` macro are wrapped.

# Example
The bodies of these plugs will all be wrapped:

```
pipeline :browser do
    plug :accepts, ["html"]
    plug :fetch_session
    plug :fetch_flash
    plug :protect_from_forgery
    plug :put_secure_browser_headers
end
```
Whereas the body of `MyApp.Plug.Authenticate` in the code below will not be wrapped:

```
scope "/", MyApp do
    pipe_through [:browser, MyApp.Plug.Authenticate]
    ...
end
```

# Workaround

Always only list pipelines in pipe_through:

```
pipeline :browser do
    plug :accepts, ["html"]
    plug :fetch_session
    plug :fetch_flash
    plug :protect_from_forgery
    plug :put_secure_browser_headers
end

pipeline :authenticate do
    plug: MyApp.Plug.Authenticate
end

scope "/", MyApp do
    pipe_through [:browser, :authenticate]
    ...
end
```

Of course, in this example the two pipelines can be combined, but that is besides the point.

**Comment by josevalim** [comment]
Yes, we could probably document it together but I don't expect us to wrap each plug in a try/catch.

**Comment by palm86** [comment]
It's not so much about wrapping everything than it is about being consistent. If the plugs in the pipeline are wrapped then I also expect the plugs in a pipe_through to be wrapped. But I can see that doing so would doubly wrap plugs that were already wrapped (as part of a pipeline). Perhaps then it should be strongly discouraged to list "raw" plugs (with raw I mean ones not in a pipeline) in the pipe_through macro.

**Comment by josevalim** [comment]
The plugs in the pipeline are wrapped because the pipeline ends up defining a new plug that wraps all of them. So the wrapping happens because of the `pipeline` and not because of `pipe_through`.

**Comment by palm86** [comment]
That is correct, yes, but nowhere in the docs. I'll make a PR to highlight that in the routing docs.

**Comment by palm86** [comment]
Just a note for posterity.

The plugs in a pipeline are all wrapped together in a single try/catch block that reraises any error that is encountered with `Plug.Conn.WrapperError.reraise(conn, kind, reason)` . That means that the `conn` that is used in the reraise doesn't reflect any of the transformations that the plugs in the pipeline could have applied to copies of the `conn`. So with this pipeline:

```
pipeline :api do
    plug :accepts, ["json"]
    plug MyApp.Plug.Authentication
end
```

any error raised in `MyApp.Plug.Authentication` will be wrapped in WrapperError, but the Phoenix error renderer will not know that the preferred format is `"json"`, because it only has access to a stale `conn` (a `conn` from before the `:accepts` plug was invoked) from the WrapperError and will therefore pick the first default from your config:

```
config :my_app, MyApp.Endpoint,
  ...
  render_errors: [accepts: ~w(html json)],
```

which in this case amounts to HTML.

In conclusion then, if the transformations applied to `conn` in a pipeline are meaningful in terms of your error response (if you rely on raising errors with a `plug_status` as opposed to explicitly calling some error view) then you have to put them in their own pipelines or wrap their bodies with the appropriate try/catch yourself.

It is not really an option for `Phoenix.Router` to wrap the plugs in a pipeline individually, because it relies on `Plug.Builder` to compile the individual plugs into a new plug.

**Comment by josevalim** [comment]
Yes, precisely. That's why the error renderer also needs to perform content negotiation... because it can be invoked before any content negotiation is performed. If you only serve APIs, then it is recommended to configure render_errors accordingly. :)

**Comment by palm86** [comment]
I'm serving both HTML and JSON, so I'm putting at least the `accepts` plug in its own pipeline. :) Thanks for the feedback.

**Comment by chrismccord** [comment]
If you have ideas where we can improve docs here, please send a PR. Thanks!
